jerry-core/ecma/operations/ecma-proxy-object.c

```diff
@@ -672,9 +672,175 @@ ecma_proxy_object_get_own_property_descriptor (ecma_object_t *obj_p, /**< proxy
+
+  /* 9. */
+  ecma_value_t trap_result = ecma_op_function_call (func_obj_p, handler, args, 2);
+  ecma_deref_object (func_obj_p);
+
+  /* 10. */
+  if (ECMA_IS_VALUE_ERROR (trap_result))
+  {
+    return trap_result;
+  }
+
+  /* 11. */
+  if (!ecma_is_value_object (trap_result) && !ecma_is_value_undefined (trap_result))
+  {
+    ecma_free_value (trap_result);
+    return ecma_raise_type_error (ECMA_ERR_MSG ("Trap is not object nor undefined."));
+  }
+
+  /* 12. */
+  ecma_property_descriptor_t target_desc;
+  ecma_value_t target_status = ecma_op_object_get_own_property_descriptor (target_obj_p, prop_name_p, &target_desc);
+
+  /* 13. */
+  if (ECMA_IS_VALUE_ERROR (target_status))
+  {
+    ecma_free_value (trap_result);
+    return target_status;
+  }
+
+  /* 14. */
+  if (ecma_is_value_undefined (trap_result))
+  {
+    /* .a */
+    if (ecma_is_value_false (target_status))
+    {
+      return ECMA_VALUE_UNDEFINED;
+    }
+    /* .b */
+    if (!(target_desc.flags & ECMA_PROP_IS_CONFIGURABLE))
+    {
+      ecma_free_property_descriptor (&target_desc);
+      return ecma_raise_type_error (ECMA_ERR_MSG ("Given property is a non-configurable"
+                                                  " data property on the proxy target"));
+    }
+
+    /* .c */
+    ecma_free_property_descriptor (&target_desc);
+    ecma_value_t extensible_target = ecma_builtin_object_object_is_extensible (target_obj_p);
+
+    /* .d */
+    if (ECMA_IS_VALUE_ERROR (extensible_target))
+    {
+      return extensible_target;
+    }
+
+    /* .e */
+    JERRY_ASSERT (ecma_is_value_boolean (extensible_target));
+
+    /* .f */
+    if (ecma_is_value_false (extensible_target))
+    {
+      return ecma_raise_type_error (ECMA_ERR_MSG ("Target not extensible."));
+    }
+
+    /* .g */
+    return ECMA_VALUE_FALSE;
+  }
+
+  /* 15. */
+  ecma_value_t extensible_target = ecma_builtin_object_object_is_extensible (target_obj_p);
+
+  /* 16. */
+  if (ECMA_IS_VALUE_ERROR (extensible_target))
+  {
+    if (ecma_is_value_true (target_status))
+    {
+      ecma_free_property_descriptor (&target_desc);
+    }
+    ecma_free_value (trap_result);
+    return extensible_target;
+  }
+
+  /* 17, 19 */
+  ecma_value_t result_val = ecma_op_to_property_descriptor (trap_result, prop_desc_p);
+
+  /* 18. */
+  ecma_op_to_complete_property_descriptor (prop_desc_p);
+  ecma_free_value (trap_result);
+
+  /* 18. */
+  if (ECMA_IS_VALUE_ERROR (result_val))
+  {
+    if (ecma_is_value_true (target_status))
+    {
+      ecma_free_property_descriptor (&target_desc);
+    }
+    return result_val;
+  }
+
+  /* 20. */
+  bool is_extensible = ecma_is_value_true (extensible_target);
+
+  bool is_valid;
+  if (ecma_is_value_true (target_status))
+  {
+    is_valid = ecma_op_is_compatible_property_descriptor (prop_desc_p, &target_desc, is_extensible);
+  }
+  else
+  {
+    is_valid = ecma_op_is_compatible_property_descriptor (prop_desc_p, NULL, is_extensible);
+  }
+
+  /* 21. */
```

Comment:
Maybe we could extract a bit of common part (the `target_desc` freeing). like:
```c

bool target_has_desc = ecma_is_value_true (target_status);
bool target_is_configurable = false;

if (target_has_desc)
{
  target_is_configurable = ((target_desc.flags & ECMA_PROP_IS_CONFIGURABLE) != 0);
  ecma_free_property_descriptor (&target_desc);
}

/* 21. */
if (!is_valid)
{
  ecma_free_property_descriptor (prop_desc_p);
  return ecma_raise_type_error (ECMA_ERR_MSG ("The two descriptors are not compatible."));
}

/* 22. */
else if (!(prop_desc_p->flags & ECMA_PROP_IS_CONFIGURABLE))
{
  if (!target_has_desc || target_is_configurable)
  {
    ecma_free_property_descriptor (prop_desc_p);
    return ecma_raise_type_error (ECMA_ERR_MSG ("Not compatible."));
  }
}
```

what do you think?

Reply:
That looks much better. Thank you. :)